### ISSUE TITLE
[BUG]: Llama makes tool calling errors

### ISSUE THREAD

**Issue by jurmy24** [Tanzania-AI-Community/twiga#63]
**Describe the bug**
I have provided my LLMs (including Llama 3.1 70b Instruct and 3.1 405b Instruct) with some test tools to see how well they work with the openai api sdk. While the tools sometimes work as they should, I often get responses from the chatbot like this:

`<function=get_weather>{"city": "New York"}"</function>`

And it does not invoke the tool_call. As it doesn't actually call any tool a new request isn't made to the api and this is what is returned to the user.

**To Reproduce**
Just have a running chatbot with the tool calls available and ask it in natural language to eg. generate an exercise or search the textbook. 

**Expected behavior**
I expect it to call the tool with proper formatting as expected by the OpenAI SDK so that my tool call functions are invoked on the server side.

**Comment by jurmy24** [comment]
Here's an incorrect response formatting sample:

`{
  "content": "<function=generate_exercise>{"query": "animal grazing question for secondary school students in Tanzania"} </function>",
  "refusal": null,
  "role": "assistant",
  "function_call": null,
  "tool_calls": []
}`


Here are some comments I got from Together AI regarding this issue and how to solve it:

``` python
def catch_malformed_tool(response_text: str) -> str | dict:
    """Parse response text to extract tool call if present, otherwise return original text."""
    FUNCTION_NAME = "screenpipe_search"
    TOOL_PREFIX = f"<function={FUNCTION_NAME}>"

    if not response_text.startswith(TOOL_PREFIX):
        return response_text
    # ... (logic to parse out the tool call)

```
    
    
Another possible solution:
    
 - Use JSON mode to force structured output function calls and then parse it out and call the functions - this will make it so the LLM won't hallucinate or produce incorrectly formatted calls as above. - https://docs.together.ai/docs/json-mode Model: meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo

**Comment by jurmy24** [comment]
![image](https://github.com/user-attachments/assets/dae048a6-adc7-4825-814f-86d4a8da174e)

**Comment by wjunwei2001** [comment]
@jurmy24 I have implemented some code changes locally for this and i think it should work but i realised even before i made some changes, the responses are like this. 
![image](https://github.com/user-attachments/assets/3fd12db5-d772-438a-9075-94d4d463ac30)

Is this what you see from your side as well? I thought <function=generate_exercise> would activate the right tool call usually

**Comment by jurmy24** [comment]
Hi @wjunwei2001. Ah no that's the way in which it formats it incorrectly. If the tool calls are formatted correctly they will look like what they do in here:

https://platform.openai.com/docs/guides/function-calling

Meaning the response should look something like this:

```python
Choice(
  finish_reason='tool_calls', 
  index=0, 
  logprobs=None, 
  message=chat.completionsMessage(
      content=None, 
      role='assistant', 
      function_call=None, 
      tool_calls=[
          chat.completionsMessageToolCall(
              id='call_62136354', 
              function=Function(
                  arguments='{"order_id":"order_12345"}', 
                  name='get_delivery_date'), 
              type='function')
      ])
)
```

Where the part we're interested in is the message property which is a ChatCompletionsMessage. We typically want the content to be None and the tool_calls to be filled when it generates a tool call. What we get is that it hallucinates and puts the stuff that should normally be in tool_calls into the content parameter when sending a response, so our code does not execute the stuff within ```if initial_message.tool_calls```.

**Comment by jurmy24** [comment]
Hi @wjunwei2001 how's it going with this. Do you need some help?

**Comment by wjunwei2001** [comment]
Hi @jurmy24, so far I have gotten it to work for one tool call where the format is "<function=...." but I have not gotten it to work for the other format {"name":.... I am just extracting the relevant information out of the content but somehow it is not working for the second format yet

**Comment by jurmy24** [comment]
Have you tried using structured json mode instead?

**Comment by wjunwei2001** [comment]
Not yet actually, for that method do you know if we have to define our own schema for the json output?

**Comment by jurmy24** [comment]
Yeah I think so, here's the link to the documentation:<br><br>[https://docs.together.ai/docs/json-mode](https://docs.together.ai/docs/json-mode)<br><br>I'm not saying this is the way to go (since it seems its limited to a few LLMs) but its worth looking into.

**Comment by wjunwei2001** [comment]
Attempted to fix the tool calling bug here 
https://github.com/Tanzania-AI-Community/twiga/commit/d7cd2e05db43bd9a39624a15db05961d9eb9359c

Not sure why it is still not working at the moment

**Comment by jurmy24** [comment]
Hey @wjunwei2001, 

I'll take a look at it soon. In the meanwhile I notice you have commits on the pre-commit.yml file. Any reason for that? Typically we want to leave the `workflows` folder unchanged. 

Also, I noticed it's a bit hard to read the changelog in the files since it says the whole file has been removed and replaced with a complete new file with your changes. Normally its just the added function that shows up as new code (eg. the catch_malformed_tool function). 

If you want to get some help from Together AI, I recommend asking them in their discord server.
You can find my previous conversation with them in the #help-and-question section "Tool call request improperly formatted".

Here's their server: https://discord.gg/9Rk6sSeWEG

**Comment by wjunwei2001** [comment]
@jurmy24 sure i can have a call with you on the 14th! Regarding the commits for the pre-commit.yml file, I am also not sure why that is there but I am guessing I probably did something wrong when trying to follow the commit format with the pre-commit and ended up committing some extra changes as well. Would explain the whole file being removed and replaced.

**Comment by jurmy24** [comment]
🤙
